File: 61-rotate-list/61-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def length(self,head):
        count = 0
        cur = head
        while cur:
            count += 1
            cur = cur.next
        return count
    
    def reverse(self,head):
        cur = head
        pre = None
        while cur!=None:
            nex = cur.next
            cur.next = pre
            pre = cur
            cur = nex
        return pre
    
    def rotateRight(self, head: Optional[ListNode], k: int):
        if head==None or head.next==None:
            return head
        
        total = self.length(head)
        k = k%total
        
        first = second = head
        move = total-k-1
        while move:
            second = second.next
            move -= 1
        temp = second.next
        second.next = None
        second = temp
        
        newH1 = self.reverse(first)
        newH2 = self.reverse(second)
        head.next = newH2
        head = self.reverse(newH1)
        return head

File: 61-rotate-list/61-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int length(ListNode* temp){
        int cnt=0;
        while(temp){
            cnt++;
            temp = temp->next;
        }
        return cnt;
    }
    
    ListNode* rotateRight(ListNode* head, int k) {
        
        if(head==NULL || head->next==NULL)
            return head;
        
        int len = length(head);
        k = k%len;
        
        ListNode *pre = head, *tail = head;
        
        while(k){
            tail = tail->next;
            k--;
        }
        
        while(tail->next){
            pre = pre->next;
            tail = tail->next;
        }
        
        tail->next = head;
        ListNode *newHead = pre->next;
        pre->next = NULL;
        return newHead;
    }
};